### DISCUSSION TITLE
Selecting a difficulty level based on the users response. HTML Button Response

### DISCUSSION THREAD

**Discussion by CayoPigeon** [jspsych/jsPsych#3268] [Q&A] [answered]
Hey everyone! This is my first time designing and experiment, so thanks in advance for the patience and understanding.  I'm trying to do so using cognition.run and the jspsych library 7.3. I want to design a task that asks the user whether they wish to modify the difficulty level on a set of questions. However, what i have so far  doesn't pick up on the user's response and doesn't modify the difficulty level. This experiment is for my thesis and I want to finish it ASAP but I've been stuck on this problem for a month now. 

This is my code for this part of the experiment. 
var jsPsych = initJsPsych({ 
  on_finish: function() {
    jsPsych.data.displayData();
  }
});

var timeline = []; 
//var dificultadActual = 3; 
var chosenDifficulty = 1; 
var difficultyLevel = 3; 

function sumMax5 (a) { // Functions allow to add or substract a level of difficulty without passing the range of 1 through 5
  var result = a; 
  result++; 
  if(result > 5)
  {
    result = 5; 
  }
  return result; 
}

function diffMin1 (a) { 
  var result = a; 
  result--; 
  if(result < 1) {
    result = 1; 
  }
  return result; 
}
var difficultySelection = {
  type: jsPsychHtmlButtonResponse,
  stimulus: "Do you wish to modify the difficulty?",
  choices: ["Easier", "Same", "Harder"],
  prompt: '',
  required: true,
  button_html: '<button class="my-custom-button" style="background-color: darkslateblue; color: snow;175px; height: 50px; border-radius: 40px;">%choice%</button>',
  on_finish: function(){
    chosenDifficulty = difficultySelection.response;
  }
};

for(var i = 0; i < 3; i++){
  timeline.push(difficultySelection); 
  if (chosenDifficulty == 0)  {//Button tag for 'Easier' is 0 
    difficultyLevel = diffMin1(difficultyLevel); 
  }
  else if (chosenDifficulty == 2) { //Button tag for 'Harder' is 2
    difficultyLevel = sumMax5(difficultyLevel); 
  }
  
  timeline.push({
    type: jsPsychHtmlKeyboardResponse,
    stimulus: "The current level of difficulty is: " + difficultyLevel + ".",
    choices: "NO_KEYS",
    trial_duration: 1500
}); } 

jsPsych.run(timeline);

**Comment by Shaobin-Jiang** [comment, marked as answer]
You have made exactly the same mistakes new learners of jsPsych commonly make (not a mockery, just my personal opinion that the official documentation never laid enough stress on the matter).

I have been teaching my fellow students on jsPsych, and whenever I do, I would tell them this: creating a trial in jsPsych does not mean that it is run upon creation, and running a trial in jsPsych does not mean the parameters are being evaluated when the trial is run. So how have you been mistaken?

You are pushing the trials into the timeline with a loop, and you wrote down these lines:

```javascript
timeline.push(difficultySelection);
if (chosenDifficulty == 0) {//Button tag for 'Easier' is 0
    difficultyLevel = diffMin1(difficultyLevel);
} else if (chosenDifficulty == 2) { //Button tag for 'Harder' is 2
    difficultyLevel = sumMax5(difficultyLevel);
}
```

No doubt you want to adjust the difficulty level after a difficulty-selection trial. However, take a look at the first half of my saying:

> creating a trial in jsPsych does not mean that it is run upon creation

The trial is created, but it is not run until `jsPsych.run` is called. So what you have done is, create a trial, meddle with the difficulty level using the **original** value with `chosenDifficulty` while what you should be using is the value selected by the user, which apparently has not happened since the experiment has not started, repeat it three times, and run the experiment. Of course it would not work. If you want something to be executed **after** a trial, put it in the `on_finish` event of a trial:

```javascript
var difficultySelection = {
    type: jsPsychHtmlButtonResponse,
    stimulus: 'Do you wish to modify the difficulty?',
    choices: ['Easier', 'Same', 'Harder'],
    prompt: '',
    required: true,
    button_html: '%choice%',
    on_finish: function (data) {
        chosenDifficulty = data.response;
        if (chosenDifficulty == 0) {
            //Button tag for 'Easier' is 0
            difficultyLevel = diffMin1(difficultyLevel);
        } else if (chosenDifficulty == 2) {
            //Button tag for 'Harder' is 2
            difficultyLevel = sumMax5(difficultyLevel);
        }
    },
};
```

I see you have already used an `on_finish` event, but there you err again, as the data from the finished trial is not acquired from the trial object: it does not contain any data. The data object is stored somewhere else, and the easiest way is to pass a `data` argument to the `on_finish` function and use it to get the data from the last trial, see code above.

Now, back away from `on_finish`. The second half of my saying:

> running a trial in jsPsych does not mean the parameters are being evaluated when the trial is run

You have this in your loop:

```javascript
timeline.push({
    type: jsPsychHtmlKeyboardResponse,
    stimulus: "The current level of difficulty is: " + difficultyLevel + ".",
    choices: "NO_KEYS",
    trial_duration: 1500
});
```

You must have intended the trial to show the chosen difficulty level using the value modified by the user. However, the parameters of a trial is calculated when the trial object is defined, not when the experiment starts to run. And there you have it, you have created three trials which all show the initial `difficultyLevel`. What you should be doing is using a dynamic parameter, or, to put it in a simpler way, use a function to return the desired value. In that way the parameter is evaluated when the trial is actually run:

```javascript
timeline.push({
    type: jsPsychHtmlKeyboardResponse,
    stimulus: () => 'The current level of difficulty is: ' + difficultyLevel + '.',
    choices: 'NO_KEYS',
    trial_duration: 1500,
});
```

 See how I use an arrow function instead to return the value.

The complete code:

```javascript
var jsPsych = initJsPsych({
    on_finish: function () {
        jsPsych.data.displayData();
    },
});

var timeline = [];
//var dificultadActual = 3;
var chosenDifficulty = 1;
var difficultyLevel = 3;

function sumMax5(a) {
    // Functions allow to add or substract a level of difficulty without passing the range of 1 through 5
    var result = a;
    result++;
    if (result > 5) {
        result = 5;
    }
    return result;
}

function diffMin1(a) {
    var result = a;
    result--;
    if (result < 1) {
        result = 1;
    }
    return result;
}
var difficultySelection = {
    type: jsPsychHtmlButtonResponse,
    stimulus: 'Do you wish to modify the difficulty?',
    choices: ['Easier', 'Same', 'Harder'],
    prompt: '',
    required: true,
    button_html: '%choice%',
    on_finish: function (data) {
        chosenDifficulty = data.response;
        if (chosenDifficulty == 0) {
            //Button tag for 'Easier' is 0
            difficultyLevel = diffMin1(difficultyLevel);
        } else if (chosenDifficulty == 2) {
            //Button tag for 'Harder' is 2
            difficultyLevel = sumMax5(difficultyLevel);
        }
    },
};

for (var i = 0; i < 3; i++) {
    timeline.push(difficultySelection);

    timeline.push({
        type: jsPsychHtmlKeyboardResponse,
        stimulus: () => 'The current level of difficulty is: ' + difficultyLevel + '.',
        choices: 'NO_KEYS',
        trial_duration: 1500,
    });
}

jsPsych.run(timeline);
```

**Reply by CayoPigeon** [reply]
Thank you so so much! You hit the nail on the head. So what the arrow does is that it updates the value of a variable based on the current run? Should I use it any time i want to call back to the response given by the user?

**Reply by Shaobin-Jiang** [reply]
The arrow is called an arrow function. This one below:

```javascript
() => value
```

is almost the same as:

```javascript
function () {
    return value;
}
```

This approach, which is setting the parameter's value to a function that returns the desired value, is called a dynamic parameter. You can use it not only when you want the parameter to be determined by a previous response by the user, but whenever the value of parameter cannot be decided at the start of the experiment.

**Reply by CayoPigeon** [reply]
Could i ask another question? 

var jsPsych = initJsPsych({
    on_finish: function () {
        jsPsych.data.displayData(); }, });

var timeline = [];
var respuestasCorrectas = 0;
var nivelDificultad = 3; 

function sumMax5(a) {
    let result = a;
    result++;
    if (result > 5) {
        result = 5; }
    return result; }

function diffMin1(a) {
    let result = a;
    result--;
    if (result < 1) {
        result = 1; }
    return result; }

var preguntasfaciles = [
  {
    stimulus: '¿En qué año se celebraron los Juegos Olímpicos en México?',
    choices: ['1966', '1968', '1973'],
    correct: 2 // Índice de la respuesta correcta en el arreglo 'opciones'
  },
  {
    stimulus: 'En las cortes medievales, ¿quién entretenía al Rey?',
    choices: ['Bufón', 'Vasallo', 'Súbdito'],
    correct: 0
  }
];

    
var seleccionDificultad = {
  type: jsPsychHtmlButtonResponse,
  stimulus: "¿Deseas modificar la dificultad?",
  choices: ["Más fácil", "Igual", "Más difícil"],
  prompt: '',
  required: true,
  button_html: '<button class="my-custom-button" style="background-color: darkslateblue; color: snow;175px; height: 50px; border-radius: 40px;">%choice%</button>',
  on_finish: function(data) {
    if (data.response == 0) { 
    nivelDificultad = diffMin1(nivelDificultad); }
    else if (data.response == 2) { 
    nivelDificultad = sumMax5(nivelDificultad); }
    data.nivelDificultad = nivelDificultad;
    return {data, nivelDificultad};
  }
};

preguntasfaciles.forEach(function(pregunta, index) {
  var EnsayoPreguntasFaciles = {
    type: jsPsychHtmlButtonResponse, 
    stimulus: pregunta.stimulus, 
    choices: pregunta.choices, 
    prompt: '',
    trial_duration: 30000,
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: cornflowerblue; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    data: {
      stimulus: pregunta.stimulus, 
      choices: pregunta.choices,
      correct: pregunta.correct }, 
    on_finish: function(data) {
      data: {response: pregunta.response};
      if (data.response == pregunta.correct) {
        respuestasCorrectas++;
        data: {correctas: respuestasCorrectas}; } }, }; 
    timeline.push(EnsayoPreguntasFaciles); 
}); 

//timeline.push(BloqueFaciles); 
timeline.push(seleccionDificultad); 

if(()=>nivelDificultad === 5) {
  timeline.push({
    type: jsPsychHtmlKeyboardResponse,
    stimulus: 'Ya funciona',
    choices: "NO_KEYS",
    trial_duration: 3000
  }); }
else if( ()=> nivelDificultad === 3) {
  timeline.push({
    type: jsPsychHtmlKeyboardResponse,
    stimulus: 'Ya funciona en 3',
    choices: "NO_KEYS",
    trial_duration: 3000
  }); }
  
jsPsych.run(timeline);

This is my code so far. My data is saving the value of nivelDificultad for the trials, however I'm still having trouble with the dynamic parameter. How can i use correctly the arrow function in order to indicate that i want to use the value assigned after the on_finish function in the seleccionDificultad trial. For some reason i don't understand, the if condition of nivelDificultad is always evaluated as true regardless of the actual value of the variable.

**Reply by Shaobin-Jiang** [reply]
Dynamic parameter is a jsPsych feature, not one that can be used and expected to work that way in any random javascript syntax. You should be doing it like this:

```javascript
timeline.push({
    type: jsPsychHtmlKeyboardResponse,
    stimulus: function () {
        if (nivelDificultad === 5) {
            return 'Ya funciona';
        } else {
            return 'Ya funciona en 3';
        }
    },
    choices: 'NO_KEYS',
    trial_duration: 3000,
});
```

**Reply by CayoPigeon** [reply]
Hey Shaobin! Would you mind helping me with something else? I've incorporated the dynamic parameters as you suggested and they worked so far. However im having trouble now with something different yet similar. 

var jsPsych = initJsPsych({
    on_finish: function() {
        jsPsych.data.displayData(); }
});

////////////////////////////
var timeline = [];
var respuestasCorrectas = 0;
var nivelDificultad = 3; 
var recorrido_mf = 0; 
var recorrido_f = 0; 
var recorrido_m = 0; 

function sumMax5(a) {
  let result = a;
  result++;
  if (result > 5) {
    result = 5; }
  return result;
}
  
function diffMin1(a) {
  let result = a;
  result--;
  if (result < 1) {
    result = 1; }
  return result;
} 

var EnsayosPreguntasMuyFaciles1 = []; 
preguntasmuyfaciles1.forEach(function(pregunta){
  var PreguntasDeEnsayo = { 
    type: jsPsychHtmlButtonResponse, 
    stimulus: pregunta.stimulus, 
    choices: pregunta.choices, 
    prompt: '',
    trial_duration: 30000,
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: cornflowerblue; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    data: {
      stimulus: pregunta.stimulus, 
      choices: pregunta.choices,
      correct: pregunta.correct }, 
    on_finish: function(data) {
      data: {response: pregunta.response};
      if (data.response == pregunta.correct) {
        respuestasCorrectas++;
        data: {correctas: respuestasCorrectas}; } 
      recorrido_mf ++; 
      return recorrido_mf; 
    }, };
  EnsayosPreguntasMuyFaciles1.push(PreguntasDeEnsayo); 
});

var EnsayosPreguntasMuyFaciles2 = []; 
preguntasmuyfaciles2.forEach(function(pregunta){
  var PreguntasDeEnsayo = { 
    type: jsPsychHtmlButtonResponse, 
    stimulus: pregunta.stimulus, 
    choices: pregunta.choices, 
    prompt: '',
    trial_duration: 30000,
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: cornflowerblue; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    data: {
      stimulus: pregunta.stimulus, 
      choices: pregunta.choices,
      correct: pregunta.correct }, 
    on_finish: function(data) {
      data: {response: pregunta.response};
      if (data.response == pregunta.correct) {
        respuestasCorrectas++;
        data: {correctas: respuestasCorrectas}; } 
      recorrido_mf ++; 
      return recorrido_mf; 
    }, };
  EnsayosPreguntasMuyFaciles2.push(PreguntasDeEnsayo); 
});

var EnsayosPreguntasFaciles1 = []; 
preguntasfaciles1.forEach(function(pregunta){
  var PreguntasDeEnsayo = { 
    type: jsPsychHtmlButtonResponse, 
    stimulus: pregunta.stimulus, 
    choices: pregunta.choices, 
    prompt: '',
    trial_duration: 30000,
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: darkolivegreen; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    data: {
      stimulus: pregunta.stimulus, 
      choices: pregunta.choices,
      correct: pregunta.correct }, 
    on_finish: function(data) {
      data: {response: pregunta.response};
      if (data.response == pregunta.correct) {
        respuestasCorrectas++;
        data: {correctas: respuestasCorrectas}; }
      recorrido_f ++; 
      return recorrido_f; 
    }, };
  EnsayosPreguntasFaciles1.push(PreguntasDeEnsayo); 
}); 

var EnsayosPreguntasFaciles2 = []; 
preguntasfaciles2.forEach(function(pregunta){
  var PreguntasDeEnsayo = { 
    type: jsPsychHtmlButtonResponse, 
    stimulus: pregunta.stimulus, 
    choices: pregunta.choices, 
    prompt: '',
    trial_duration: 30000,
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: darkolivegreen; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    data: {
      stimulus: pregunta.stimulus, 
      choices: pregunta.choices,
      correct: pregunta.correct }, 
    on_finish: function(data) {
      data: {response: pregunta.response};
      if (data.response == pregunta.correct) {
        respuestasCorrectas++;
        data: {correctas: respuestasCorrectas}; }
      recorrido_f ++; 
      return recorrido_f; 
    }, };
  EnsayosPreguntasFaciles2.push(PreguntasDeEnsayo); 
}); 

var EnsayosPreguntasMedio1 = []; 
preguntasmedio1.forEach(function(pregunta){
  var PreguntasDeEnsayo = { 
    type: jsPsychHtmlButtonResponse, 
    stimulus: pregunta.stimulus, 
    choices: pregunta.choices, 
    prompt: '',
    trial_duration: 30000,
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: thistle; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    data: {
      stimulus: pregunta.stimulus, 
      choices: pregunta.choices,
      correct: pregunta.correct }, 
    on_finish: function(data) {
      data: {response: pregunta.response};
      if (data.response == pregunta.correct) {
        respuestasCorrectas++;
        data: {correctas: respuestasCorrectas}; }
      recorrido_m ++; 
      return recorrido_m; 
    }, };
  EnsayosPreguntasMedio1.push(PreguntasDeEnsayo); 
});  

var EnsayosPreguntasMedio2 = []; 
preguntasmedio2.forEach(function(pregunta){
  var PreguntasDeEnsayo = { 
    type: jsPsychHtmlButtonResponse, 
    stimulus: pregunta.stimulus, 
    choices: pregunta.choices, 
    prompt: '',
    trial_duration: 30000,
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: thistle; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    data: {
      stimulus: pregunta.stimulus, 
      choices: pregunta.choices,
      correct: pregunta.correct }, 
    on_finish: function(data) {
      data: {response: pregunta.response};
      if (data.response == pregunta.correct) {
        respuestasCorrectas++;
        data: {correctas: respuestasCorrectas}; }
      recorrido_m ++; 
      return recorrido_m; 
    }, };
  EnsayosPreguntasMedio2.push(PreguntasDeEnsayo); 
}); 

var afecto = {
  type: jsPsychSurveyMultiChoice,
  questions: [{
    prompt: "Mi desempeño hasta ahora me hace sentir...", 
    name: 'AfectoDesempeño', 
    options: ['Bastante negativo', 'Negativamente', 'Ligeramente negativo', 'Ligeramente positivo', 'Positivo', 'Bastante positivo'], 
    required: true, 
    horizontal: true }]
};

var retroalimentacion = {
  type: jsPsychHtmlKeyboardResponse, 
  stimulus: function() {
    var feedback = "";
    if (respuestasCorrectas === 5) {
      feedback = "Muy bien";
    } else if (respuestasCorrectas === 4 || respuestasCorrectas === 3){
      feedback = "Bien";
    } else if (respuestasCorrectas <= 2) { 
      feedback = "Puede mejorar"; }
    return `<div style="font-size: 20px;"><p>${feedback}</p><p>Respuestas correctas: ${respuestasCorrectas} de ${10 - 5}.</p></div>`;
  },
  choices: "NO_KEYS",
  trial_duration: 3000, // Ajusta esta duración según lo necesario
  on_finish: function () {
    respuestasCorrectas = 0;
    return respuestasCorrectas; }
};

var seleccionDificultad = {
  type: jsPsychHtmlButtonResponse,
  stimulus: "¿Deseas modificar la dificultad?",
  choices: ["Más fácil", "Igual", "Más difícil"],
  prompt: '',
  required: true,
  button_html: '<button class="my-custom-button" style="background-color: darkslateblue; color: snow;175px; height: 50px; border-radius: 40px;">%choice%</button>',
  on_finish: function(data) {
    if (data.response == 0) { 
      nivelDificultad = diffMin1(nivelDificultad); }
    else if (data.response == 2) { 
      nivelDificultad = sumMax5(nivelDificultad); }
    data.nivelDificultad = nivelDificultad;
    return {data, nivelDificultad}; }
}; 

var SegundaMitadBloque = {
  timeline:[afecto, retroalimentacion, seleccionDificultad]
}; 


var TimelinePreguntasMuyFaciles = { 
  timeline:function () {
    if (recorrido_mf < 5) {
      return [EnsayosPreguntasMuyFaciles1[0], EnsayosPreguntasMuyFaciles1[1], EnsayosPreguntasMuyFaciles1[2], EnsayosPreguntasMuyFaciles1[3], EnsayosPreguntasMuyFaciles1[4]]; 
    }
    else if (recorrido_mf > 4){ 
      return [EnsayosPreguntasMuyFaciles2[0], EnsayosPreguntasMuyFaciles2[1], EnsayosPreguntasMuyFaciles2[2], EnsayosPreguntasMuyFaciles2[3], EnsayosPreguntasMuyFaciles2[4]];
    }
  },
  conditional_function: function () {
    if (nivelDificultad < 3) { 
      return true; 
    }
    else {
      return false; }
  }
}; 

var TimelinePreguntasFaciles = {
  timeline:function () {
    if(recorrido_f < 5) {
      return [EnsayosPreguntasFaciles1[0], EnsayosPreguntasFaciles1[1], EnsayosPreguntasFaciles1[2], EnsayosPreguntasFaciles1[3], EnsayosPreguntasFaciles1[4]];
    }
    else if(recorrido_f > 4){
     return [EnsayosPreguntasFaciles2[0], EnsayosPreguntasFaciles2[1], EnsayosPreguntasFaciles2[2], EnsayosPreguntasFaciles2[3], EnsayosPreguntasFaciles2[4]];
    }
  },
  conditional_function: function () {
    if (nivelDificultad == 3) { 
      return true; }
    else {
      return false; }
  }
};

var TimelinePreguntasMedio = {
  timeline:function () {
    if(recorrido_m < 5) {
      return [EnsayosPreguntasMedio1[0], EnsayosPreguntasMedio1[1], EnsayosPreguntasMedio1[2], EnsayosPreguntasMedio1[3], EnsayosPreguntasMedio1[4]];
    }
    else { 
      return [EnsayosPreguntasMedio2[0], EnsayosPreguntasMedio2[1], EnsayosPreguntasMedio2[2], EnsayosPreguntasMedio2[3], EnsayosPreguntasMedio2[4]];
    }  
  },
  conditional_function: function () {
    if (nivelDificultad > 3) {
      return true; } 
    else {
      return false; }
  }
};

for (var ii = 0; ii < 5; ii++) {
  timeline.push(TimelinePreguntasMuyFaciles, TimelinePreguntasFaciles, TimelinePreguntasMedio);
  timeline.push(SegundaMitadBloque); 
}

jsPsych.run(timeline);

The first timeline.push inside the for cycle isn't running anything. I'm thinking maybe the problem is with the functions declared within the timeline parameters of the last three variables i declared: TimelinePreguntasMuyFaciles, TimelinePreguntasFaciles and TimelinePreguntasMedio. Do you see anything i could be doing wrong? 
My idea is to have a sort of counter which are the variables recorrido_mf, recorrido_f, and recorrido_m to keep track of what ive used already. Perhaps it has something to do with those counters? I have them because the idea of the experiment is to ask a set of questions and the user is supposed to be able to select the level of difficulty of those questions. I'm struggling now with how to avoid the repetition of questions. The set of questions are uploaded on a js file in cognition.run. 
What do you think of what I've done so far? Am i using correctly the dynamic parameters?

**Reply by Shaobin-Jiang** [reply]
The `timeline` cannot be used as a dynamic parameter.

**Reply by CayoPigeon** [reply]
Thank you for telling me that. I didn't know that. What do you think about this? 
var jsPsych = initJsPsych({
    on_finish: function() {
        jsPsych.data.displayData(); }
});

var timeline = [];
var respuestasCorrectas = 0;
var nivelDificultad = 3; 
var recorrido_mf = 0; 
var recorrido_f = 0; 
var recorrido_m = 0; 

function sumMax5(a) {
  let result = a;
  result++;
  if (result > 5) {
    result = 5; }
  return result;
}
  
function diffMin1(a) {
  let result = a;
  result--;
  if (result < 1) {
    result = 1; }
  return result;
} 

function Contador (avance, ii){
  var c = ii; 
  if(avance > 4) { 
    c + 5; }
  return c; 
}

var EnsayosMuyFacilesPrueba = []; 
for (var i = 0; i < preguntasmuyfaciles.length; i++) {
  var PreguntaMFPrueba = {
    type: jsPsychHtmlButtonResponse, 
    stimulus: preguntasmuyfaciles[Contador(recorrido_mf, i)].stimulus,
    choices: preguntasmuyfaciles[Contador(recorrido_mf, i)].choices,
    prompt: '', 
    trial_duration: 30000, 
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: cornflowerblue; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    on_finish: function(data) {
       if (data.response == preguntasmuyfaciles[Contador(recorrido_mf, i)].correct) {
        respuestasCorrectas++; 
        data.correctas = respuestasCorrectas; 
      }
      recorrido_mf++;
    }
  }; 
  EnsayosMuyFacilesPrueba.push(PreguntaMFPrueba); 
}

var EnsayosFacilesPrueba = []; 
for (var i = 0; i < preguntasfaciles.length; i++) {
  var PreguntaFPrueba = {
    type: jsPsychHtmlButtonResponse, 
    stimulus: preguntasfaciles[Contador(recorrido_f, i)].stimulus,
    choices: preguntasfaciles[Contador(recorrido_f, i)].choices,
    prompt: '', 
    trial_duration: 30000, 
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: darkolivegreen; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    on_finish: function(data) {
      if (data.response == preguntasfaciles[Contador(recorrido_f, i)].correct) {
        respuestasCorrectas++; 
        data.correctas = respuestasCorrectas; 
      }
      recorrido_f++; 
    }
  }; 
  EnsayosFacilesPrueba.push(PreguntaFPrueba); 
}

var EnsayosMedioPrueba = []; 
for (var i = 0; i < preguntasmedio.length; i++) {
  var PreguntaMPrueba = {
    type: jsPsychHtmlButtonResponse, 
    stimulus: preguntasmedio[Contador(recorrido_m, i)].stimulus,
    choices: preguntasmedio[Contador(recorrido_m, i)].choices,
    prompt: '', 
    trial_duration: 30000, 
    required: true, 
    button_html: '<button class="my-custom-button" style="background-color: thistle; color: ghostwhite;175px; height: 50px; border-radius: 40px;">%choice%</button>', 
    on_finish: function(data) {
      if (data.response == preguntasmedio[Contador(recorrido_m, i)].correct) {
        respuestasCorrectas++; 
        data.correctas = respuestasCorrectas; 
      }
      recorrido_m++; 
    }
  }; 
  EnsayosMedioPrueba.push(PreguntaMPrueba); 
}

var afecto = {
  type: jsPsychSurveyMultiChoice,
  questions: [{
    prompt: "Mi desempeño hasta ahora me hace sentir...", 
    name: 'AfectoDesempeño', 
    options: ['Bastante negativo', 'Negativamente', 'Ligeramente negativo', 'Ligeramente positivo', 'Positivo', 'Bastante positivo'], 
    required: true, 
    horizontal: true }]
};

var retroalimentacion = {
  type: jsPsychHtmlKeyboardResponse, 
  stimulus: function() {
    var feedback = "";
    if (respuestasCorrectas === 5) {
      feedback = "Muy bien";
    } else if (respuestasCorrectas === 4 || respuestasCorrectas === 3){
      feedback = "Bien";
    } else if (respuestasCorrectas <= 2) { 
      feedback = "Puede mejorar"; }
    return `<div style="font-size: 20px;"><p>${feedback}</p><p>Respuestas correctas: ${respuestasCorrectas} de ${10 - 5}.</p></div>`;
  },
  choices: "NO_KEYS",
  trial_duration: 3000, // Ajusta esta duración según lo necesario
  on_finish: function () {
    respuestasCorrectas = 0;
    return respuestasCorrectas; }
};

var seleccionDificultad = {
  type: jsPsychHtmlButtonResponse,
  stimulus: "¿Deseas modificar la dificultad?",
  choices: ["Más fácil", "Igual", "Más difícil"],
  prompt: '',
  required: true,
  button_html: '<button class="my-custom-button" style="background-color: darkslateblue; color: snow;175px; height: 50px; border-radius: 40px;">%choice%</button>',
  on_finish: function(data) {
    if (data.response == 0) { 
      nivelDificultad = diffMin1(nivelDificultad); }
    else if (data.response == 2) { 
      nivelDificultad = sumMax5(nivelDificultad); }
    data.nivelDificultad = nivelDificultad;
    return {data, nivelDificultad}; }
}; 

var SegundaMitadBloque = {
  timeline:[afecto, retroalimentacion, seleccionDificultad]
}; 

var TimelineMuyFacilesPrueba = {
  timeline:[EnsayosMuyFacilesPrueba[0], EnsayosMuyFacilesPrueba[1], EnsayosMuyFacilesPrueba[2], EnsayosMuyFacilesPrueba[3], EnsayosMuyFacilesPrueba[4]],
  conditional_function: function() {
    if (nivelDificultad < 3) { 
      return true; 
    }
    else {
      return false; }
  }
}; 

var TimelineFacilesPrueba = {
  timeline:[EnsayosFacilesPrueba[0], EnsayosFacilesPrueba[1], EnsayosFacilesPrueba[2], EnsayosFacilesPrueba[3], EnsayosFacilesPrueba[4]],
  conditional_function: function() {
    if (nivelDificultad == 3) { 
      return true; 
    }
    else {
      return false; }
  }
}; 

var TimelineMedioPrueba = {
  timeline:[EnsayosMedioPrueba[0], EnsayosMedioPrueba[1], EnsayosMedioPrueba[2], EnsayosMedioPrueba[3], EnsayosMedioPrueba[4]],
  conditional_function: function() {
    if (nivelDificultad > 3) { 
      return true; 
    }
    else {
      return false; }
  }
}; 

var BloqueCompleto = {
  timeline:[TimelineMuyFacilesPrueba, TimelineFacilesPrueba, TimelineMedioPrueba, SegundaMitadBloque]
};

for(var ii = 0; ii < 5; ii++){
  timeline.push(BloqueCompleto); 
}
jsPsych.run(timeline);
